llvm/utils/TableGen/DynamicTableEmitter.cpp

```diff
@@ -0,0 +1,267 @@
+        return std::string(SI->getValue());
+
+      return SI->getAsString();
+    }
+    if (BitsInit *BI = dyn_cast<BitsInit>(I))
+      return "0x" + utohexstr(getAsInt(BI));
+    if (BitInit *BI = dyn_cast<BitInit>(I))
+      return BI->getValue() ? "true" : "false";
+    if (Field.IsList) {
+      if (auto LI = dyn_cast<ListInit>(I)) {
+        std::stringstream Result;
+        // Open list
+        Result << "{";
+        auto Values = LI->getValues();
+        bool IsAspect = (Field.Name == "aspects");
+        ListSeparator LS;
+        for (const auto &[Idx, Val] : enumerate(Values)) {
+          // ListSeparator only provides the StringRef() operator.
+          StringRef Separator = LS;
+          Result << Separator.str();
+
+          if (!IsAspect)
+            Result << Val->getAsString();
+          else {
+            auto Record = LI->getElementAsRecord(Idx);
+            Result << "\"" << Record->getValueAsString("Name").str() << "\"";
+          }
+        }
+        // Close list
+        Result << "}";
+        return Result.str();
+      }
+      PrintFatalError(Loc,
+                      Twine("Entry for field '") + Field.Name + "' is null");
+    }
+    PrintFatalError(Loc, Twine("invalid field type for field '") + Field.Name +
+                             "'; expected: bit, bits, string, or code");
+  }
+
+  void emitDynamicTable(const DynamicTable &Table, raw_ostream &OS);
+  void emitIfdef(Twine Guard, raw_ostream &OS);
+
+  bool parseFieldType(GenericField &Field, Init *II);
+  void collectTableEntries(DynamicTable &Table,
+                           const std::vector<Record *> &Items);
+};
+
+} // End anonymous namespace.
+
+void DynamicTableEmitter::emitIfdef(Twine Guard, raw_ostream &OS) {
+  OS << "#ifdef " << Guard.str() << "\n";
+  PreprocessorGuards.insert(Guard.str());
+}
+
+void DynamicTableEmitter::emitDynamicTable(const DynamicTable &Table,
+                                           raw_ostream &OS) {
+  emitIfdef((Twine("GET_") + Table.PreprocessorGuard + "_IMPL"), OS);
+
+  // The primary data table contains all the fields defined for this map.
+  OS << "std::map<std::string, " << Table.CppTypeName << "> " << Table.Name
+     << " = {\n";
+  // Iterate over the key-value pairs the dynamic table will contain.
+  for (unsigned i = 0; i < Table.Entries.size(); ++i) {
+    Record *Entry = Table.Entries[i];
+    // Open key-value pair
+    OS << "  { ";
+
+    ListSeparator MapElemSeparator;
+    ListSeparator TargetInfoElemSeparator;
+    // Iterate over the different fields of each entry. First field is the key,
+    // the rest of fields are part of the value.
+    for (const auto &[Idx, Field] : enumerate(Table.Fields)) {
+      bool IsKey = (Idx == 0);
+      std::string val = primaryRepresentation(Table.Locs[0], Field,
+                                              Entry->getValueInit(Field.Name));
+      if (!IsKey)
+        OS << TargetInfoElemSeparator << val;
+      else {
```

Comment:
Nit: I prefer having either all branches having bodies or none of them having.
```suggestion
      if (!IsKey) {
        OS << TargetInfoElemSeparator << val;
      } else {
```